### DISCUSSION TITLE
Questions about the radio, datalink and communication from nRF chip to the STM32 in Crazyflie 2.1

### DISCUSSION THREAD

**Discussion by Pelochus** [bitcraze/discussions#1256] [Q&A] [answered]
I am currently trying to use Paparazzi UAV with the Crazyflie 2.1, since they support it. Both I and the Paparazzi team have been trying to debug what is going on with my Crazyflie that does not send data back to the Crazyradio. 

After fixing some problems here and there (LEDs, RTOS inside the firmware, problems related to Paparazzi or the software installed on my PC...) we have reached the conclusion that the problem is most likely within the communication within the STM32 chip and the nRF chip OR, as a secondary possible problem, with the communication between the nRF chip and the Crazyradio 2.0

## Explaining the issue
Paparazzi uses a Python script that uses ```cflib``` library to communicate with the Crazyradio 2.0. This script also sends the data received from the radio to the ```ivybus``` which is the layer of Paparazzi responsible for datalink between ground station (my PC) and the UAVs (Crazyflie). Paparazzi then displays this data or sends data back.

My issue is very simple: the radio does in fact receive packets from the Crazyflie but they are always empty (no data inside them), so Paparazzi ends up discarding them. Crazyradio 2.0 has a green LED on and flashing red LED when the Crazyflie 2.1 is on, so **there is connection between the nRF chip and the radio**. The problem seems to be between the STM32 and the nRF chip.

## What I tried to solve it
Is important to know that the **nRF firmware used is the same than in the official Crazyflie 2.1 firmware**. Knowing this, I tried flashing an older official firmware and then flashing over Paparazzi's firmware again. This actually solved one LED that did not turn on previously (still wondering why this happened) but same behaviour with the radio: empty packets sent constantly to the radio. 

As a last resort, I asked the developer who made the implementation of Paparazzi for Crazyflie, used his firmware binaries (which, according to him, worked for his Crazyflie) and flashed them to my Crazyflie and still had no data communication, only empty packets. We both have **Crazyflie 2.1 Rev B1**, but it is **important to know that he has the original Crazyradio PA**. It is also very interesting that he never flashed any official firmware update, so perhaps he has a different nRF firmware from factory that works with Paparazzi (he bought the Crazyflie in 2019).

## What made me believe that the problem is not Paparazzi's fault
Since I used the binaries from the developer who has a working Crazyflie with Paparazzi firmware and then flashing older firmwares from Bitcraze and flashing again Paparazzi solved that one LED working, I am suspecting that the problem is in the nRF firmware or something similar, related to it (communication between STM32 and nRF and/or the radio)

**All that said, my question is: What do you guys think could be the problem?** Could you explain me a bit more or reference some advanced documentation relating to the nRF, Crazyradio and communication between STM32 and nRF chip? Is there any unreleased firmware for the Crazyflie that has a different nRF firmware version that could, perhaps, have a different way of communicating with the STM32?

For more information about the issue, you can check this link, though the last 5-8 comments are the most relevant to these topic. Previous comments are related to other issues that are already solved or partially solved: 

https://github.com/paparazzi/paparazzi/issues/3252

**Comment by knmcguire** [comment]
Hi! We don't use paparazzi ourselves so it will be difficult for us to debug this properly, and we are also a bit low on time. But let me give you some links that might be helpful to dig in:

Repository documentation with more details:
- STM firmware: https://www.bitcraze.io/documentation/repository/crazyflie2-nrf-firmware/master/
-  NRF51 firmware: https://www.bitcraze.io/documentation/repository/crazyflie2-nrf-firmware/master/
- CFlib: https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/
- Crazyradio 2: https://www.bitcraze.io/documentation/repository/crazyradio2-firmware/main/
These can also be found through here: https://www.bitcraze.io/documentation/repository/

I'm not sure if related but we did had to change some things on the NRF firmware itself of a latest release: https://www.bitcraze.io/2024/02/2024-2-firmware-release-with-new-bluetooth-stack/. However you did mention that you also downgraded the Crazyflie as well and you still had the same issue. 

I'll forward this to some of my colleagues to see if they have any more tips

**Reply by Pelochus** [reply]
Thank you very much, I'll check those links out soon.

**Comment by Pelochus** [comment]
Okay, so I just found out that the script from the Paparazzi team does receive empty packets from the official firmware too. Crazyflie client does receive telemetry, so that suggests that the problem is very likely within the python script Paparazzi uses for the communication from the radio. 

Since the difference between my setup and the setup from the guy from the Paparazzi team, is the Crazyradio (me: 2.0 (not working), him: PA (working)) I believe the problem is in either the Python script or ```cflib```

Most likely, I believe there are some differences between Crazyradio 2.0 and Crazyradio PA at the ```cflib``` level, so I just want you guys to confirm that those differences exist and exactly what are them and what I need to change in the script.

If you don't mind, I'll link here the Paparazzi script. Is essentially using ```cflib``` so it may be very easy to read for you guys at the Bitcraze team and soon discover what should be changed from PA to 2.0.

https://github.com/paparazzi/paparazzi/blob/master/sw/ground_segment/python/bitcraze/crazyradio2ivy.py

Note: In my case I had to change ```CRTP_PORT_PPRZLINK = 9``` to ```CRTP_PORT_PPRZLINK = 15```

**Comment by Pelochus** [comment, marked as answer]
Fixed! After all I changed one variable from the Paparazzi script to an incorrect value (even though I tried with the correct value before... anyway)

Seems like it is also necessary for Paparazzi to work to go back to older firmwares too before flashing Paparazzi over. I'm using 2019.9.

**Reply by knmcguire** [reply]
Thanks for updating us! and happy you found some sort of fix.
